### PR TITLE
add functionality to push best models to the hub during training

### PR DESCRIPTION
**Pull request by Bearnardd**
### Details

* [x] update `PPOConfig` to support arguments for saving models during training
* [x] update 'PPOTrainer` to support saving models during training
* [x]  add test coverage for the changes

### Issue
[issue-255](https://github.com/lvwerra/trl/issues/255)

### CODE REVIEW

File: tests/test_ppo_trainer.py [hunk 1]
```diff
@@ -875,3 +875,42 @@ def test_generation(self):
...
         self.assertEqual(generations_single, generations_batched)
+
+    @unittest.skip("Fix by either patching `whomai()` to work in the staging endpoint or use a dummy prod user.")
+    def test_push_to_hub_if_best_reward(self):
+        REPO_NAME = "test-ppo-trainer"
+        repo_id = f"{CI_HUB_USER}/{REPO_NAME}"
+
+        dummy_dataset = self._init_dummy_dataset()
+
+        # push_to_hub_if_best_kwargs = {"repo_id": repo_id, "token": self._token, "api_endpoint": CI_HUB_ENDPOINT}
```
Comment (younesbelkada) [inline tests/test_ppo_trainer.py:886]:
```suggestion
```

**Review by younesbelkada — COMMENTED** [review]
This looks great to me! I left a single comment
Thanks a lot for working on adding this cool feature!

File: tests/test_ppo_trainer.py (hunk 1 above)
Comment (Bearnardd) [reply tests/test_ppo_trainer.py]: @younesbelkada  done :)

File: trl/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -301,6 +301,15 @@ def __init__(
...
         self.current_step = 0
 
+        # init variables for pushing model to hub
+        if config.push_to_hub_if_best_kwargs:
+            assert (
+                "repo_id" in config.push_to_hub_if_best_kwargs
+            ), "You have to specify repo_id in order to push the model to the hub!"
+            self.push_to_hub_kwargs = config.push_to_hub_if_best_kwargs
+            self.compare_step = 0
+            self.highest_reward = 0
```
Comment (natolambert) [inline trl/trainer/ppo_trainer.py:311]:
```suggestion
            self.highest_reward = torch.tensor(-float('inf'))
```

File: tests/test_ppo_trainer.py [hunk 3]
```diff
@@ -875,3 +875,38 @@ def test_generation(self):
...
+        dummy_dataset = self._init_dummy_dataset()
+
+        push_to_hub_if_best_kwargs = {"repo_id": repo_id}
+
+        ppo_config = PPOConfig(
+            batch_size=2,
+            mini_batch_size=1,
+            log_with=None,
+            push_to_hub_if_best_kwargs=push_to_hub_if_best_kwargs,
+            steps_beetwen_check=1,
```
Comment (natolambert) [inline tests/test_ppo_trainer.py:893]:
```suggestion
            compare_steps =1,
```
Fix typo and match your variable name :)

File: trl/trainer/ppo_trainer.py (hunk 2 above)
Comment (natolambert) [reply trl/trainer/ppo_trainer.py:311]: Need this in case rewards are negative.

**Review by natolambert — COMMENTED** [review]
Small nits / fixes from me.

File: trl/trainer/ppo_trainer.py (hunk 2 above)
Comment (Bearnardd) [reply trl/trainer/ppo_trainer.py:311]: @natolambert great catch, thanks!

File: trl/trainer/ppo_trainer.py [hunk 4]
```diff
@@ -301,6 +301,15 @@ def __init__(
         # init the current step
         self.current_step = 0
 
+        # init variables for pushing model to hub
+        if config.push_to_hub_if_best_kwargs:
+            assert (
+                "repo_id" in config.push_to_hub_if_best_kwargs
+            ), "You have to specify repo_id in order to push the model to the hub!"
```
Comment (younesbelkada) [inline trl/trainer/ppo_trainer.py:308]:
```suggestion
            if "repo_id" not in config.push_to_hub_if_best_kwargs:
                raise ValueError("You have to specify repo_id in order to push the model to the hub!")
```
Let's not use `assert` in the codebase! Why is that? I invite you to read https://snyk.io/blog/the-dangers-of-assert-in-python/ !

**Review by younesbelkada — COMMENTED** [review]
Stil LGTM! Can you run the styling checks?

```bash
make style && make quality
```

File: trl/trainer/ppo_trainer.py (hunk 4 above)
Comment (Bearnardd) [reply trl/trainer/ppo_trainer.py:308]: Oh thanks!

**Comment by Bearnardd** [comment]
@younesbelkada Done :)

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for your clean work on this!

**Review by natolambert — APPROVED** [review]
LGTM too, excited to use this!

**Pull request merged**